include/pybind11/pybind11.h

```diff
@@ -1245,6 +1245,38 @@ class module_ : public object {
         //       For Python 2, reinterpret_borrow was correct.
         return reinterpret_borrow<module_>(m);
     }
+
+    /** \rst
+        Create a new top-level embedded module.
+
+        ``def`` should point to a statically allocated module_def.
+    \endrst */
+    static module_
+    create_embedded_extension_module(const char *name, const char *doc, module_def *def) {
```

Comment:
Hmm, this a lot of code duplication for changing the default value of one argument and the generic pybind11 fail, isn't it? If you think from API/ABI design they should remain seperate, maybe they can both call a common private method. Thoughts @henryiii @rwgk ?

Reply:
Maybe this would actually be better to subclass?

Reply:
I agree with @Skylion007, thinking it will be easy to avoid the code duplication, maybe as simple as `bool is_embedded = false`? (same idea as Aaron's?)

My second concern is testing: what would it take to add a test that fails without the change here?

Reply:
This just feels like bad code smell, the MACRO we are using is already a bit hacky. Is there a way we can improve create_embedded_extension_module to automatically detect when it's being called from that macro? It seems like a rather specific and limited use case that the end user should never call directly so this method should really be hidden somewhere as an implementation detail.

Reply:
Unless there is a reason to allow embedded and non-embedded modules in the same executable / library, maybe detecting whether `embed.h` has been included, or having the user indicate if they want the modules to be embedded with a `#define PYBIND11_EMBED` are options to consider ?

Reply:
@rwgk About the testing: I added a line in `external_module` which was making the "Restart an interpreter" test case fail in a way that was consistent with the original issue (ie. having the library complain about an already registered type). I could have added another test but I didn't see the point since it would have done exactly the same thing and I needed to add that line in the external module in any case.

Reply:
Okay, here is my proposal @StarQTius:
* Factor out the implementation of the static module function into to a function in a pybind11 `detail` namespace. You can have it use the is_embedded to denote the different behavior. Have the static function call the new detail one.
* Have the macro call the new detail function with the proper bool from the macro.

That way we don't create a new static method API that users can accidentally use or that we have to mantain and the API is the same as before with minimal code duplication. (The detail namespace will indicate this is not a new part of the API).

Reply:
> bout the testing: I added a line in `external_module.cpp` which was making the "Restart an interpreter" test case fail in a way that was consistent with the original issue (ie. having the library complain about an already registered type).

I see, thanks! I saw that line but didn't make the connection. Could you please add a comment, maybe one or two lines to explain in a nutshell, plus reference to this PR? So that it will be quick for anyone looking at the test code in the future, which no specific context, to make the connection.

Reply:
As you asked, I moved the extension module creation implementation detail into a non-API function. I had to add forward declarations to make it works though. I also added comments to clarify how local internals are tested, and a reference to the PR.